File: 3425.longest-special-path.py
#
# @lc app=leetcode id=3425 lang=python3
#
# [3425] Longest Special Path
#

from typing import List
import sys
from collections import defaultdict

# @lc code=start
class Solution:
    def longestSpecialPath(self, edges: List[List[int]], nums: List[int]) -> List[int]:
        sys.setrecursionlimit(10**5 + 10)
        n = len(nums)
        adj = [[] for _ in range(n)]
        for u, v, w in edges:
            adj[u].append((v, w))
            adj[v].append((u, w))
        
        self.max_len_here = [0] * n
        self.nodes_here = [1] * n
        self.left = 0
        self.path = []
        self.prefix = []
        self.pos_lists = defaultdict(list)
        
        def dfs(curr: int, par: int, edge_w: int) -> None:
            self.path.append(curr)
            if self.prefix:
                self.prefix.append(self.prefix[-1] + edge_w)
            else:
                self.prefix.append(0)
            idx = len(self.path) - 1
            val = nums[curr]
            
            new_left = self.left
            if val in self.pos_lists and self.pos_lists[val]:
                prev_idx = self.pos_lists[val][-1]
                if prev_idx >= new_left:
                    new_left = prev_idx + 1
            
            old_left = self.left
            self.left = new_left
            
            self.pos_lists[val].append(idx)
            
            this_len = self.prefix[-1] - self.prefix[self.left]
            this_nodes = idx - self.left + 1
            self.max_len_here[curr] = this_len
            self.nodes_here[curr] = this_nodes
            
            for neigh, w in adj[curr]:
                if neigh != par:
                    dfs(neigh, curr, w)
            
            self.pos_lists[val].pop()
            if not self.pos_lists[val]:
                del self.pos_lists[val]
            self.left = old_left
            
            self.prefix.pop()
            self.path.pop()
        
        dfs(0, -1, 0)
        
        global_max = max(self.max_len_here)
        min_nodes = min(self.nodes_here[i] for i in range(n) if self.max_len_here[i] == global_max)
        
        return [global_max, min_nodes]
        
# @lc code=end

File: 3425.longest-special-path.cpp
#
# @lc app=leetcode id=3425 lang=cpp
#
# [3425] Longest Special Path
#

# @lc code=start
class Solution {
public:
    vector<int> longestSpecialPath(vector<vector<int>>& edges, vector<int>& nums) {
        int n = (int)nums.size();
        vector<vector<pair<int,int>>> g(n);
        g.assign(n, {});
        for (auto &e : edges) {
            int u = e[0], v = e[1], w = e[2];
            g[u].push_back({v, w});
            g[v].push_back({u, w});
        }

        int maxVal = 50001; // nums[i] <= 5e4
        vector<int> last(maxVal, -1);

        struct Frame {
            int u, p;
            int idx;
            int L_before;
            int L_after;
            int prevLast;
            long long dist;
        };

        long long bestLen = 0;
        int bestNodes = 1;

        vector<long long> pathDist; // distance from root to node at each stack position
        pathDist.reserve(n);

        vector<Frame> st;
        st.reserve(n);

        auto enterNode = [&](int u, int p, long long dist, int L_before) {
            int pos = (int)pathDist.size();
            int v = nums[u];
            int prevLast = last[v];
            int L_after = L_before;
            if (prevLast != -1) L_after = max(L_after, prevLast + 1);

            last[v] = pos;
            pathDist.push_back(dist);

            long long len = dist - pathDist[L_after];
            int nodesCnt = pos - L_after + 1;
            if (len > bestLen) {
                bestLen = len;
                bestNodes = nodesCnt;
            } else if (len == bestLen) {
                bestNodes = min(bestNodes, nodesCnt);
            }

            st.push_back(Frame{u, p, 0, L_before, L_after, prevLast, dist});
        };

        enterNode(0, -1, 0LL, 0);

        while (!st.empty()) {
            Frame &f = st.back();
            int u = f.u;

            if (f.idx < (int)g[u].size()) {
                auto [v, w] = g[u][f.idx++];
                if (v == f.p) continue;
                enterNode(v, u, f.dist + (long long)w, f.L_after);
            } else {
                // exit node: restore state
                int val = nums[u];
                last[val] = f.prevLast;
                pathDist.pop_back();
                st.pop_back();
            }
        }

        return { (int)bestLen, bestNodes };
    }
};
# @lc code=end
